src/core/net/srp_client.cpp

```diff
@@ -0,0 +1,1559 @@
+    if (header.GetZoneRecordCount() != 0)
+    {
+        VerifyOrExit(header.GetZoneRecordCount() == 1, error = OT_ERROR_PARSE);
+        SuccessOrExit(error = Dns::Name::ParseName(aMessage, offset));
+        VerifyOrExit(offset + sizeof(Dns::Zone) <= aMessage.GetLength(), error = OT_ERROR_PARSE);
+        offset += sizeof(Dns::Zone);
+    }
+
+    // Check for Update Lease OPT RR. This determines the lease
+    // interval accepted by server. If not present, then use the
+    // transmitted lease interval from the update request message.
+
+    mAcceptedLeaseInterval = mLeaseInterval;
+    recordCount =
+        header.GetPrerequisiteRecordCount() + header.GetUpdateRecordCount() + header.GetAdditionalRecordCount();
+
+    while (recordCount > 0)
+    {
+        uint16_t            startOffset = offset;
+        Dns::ResourceRecord rr;
+
+        SuccessOrExit(error = ReadResourceRecord(aMessage, offset, rr));
+        recordCount--;
+
+        if (rr.GetType() == Dns::ResourceRecord::kTypeOpt)
+        {
+            SuccessOrExit(error = ProcessOptRecord(aMessage, startOffset, static_cast<Dns::OptRecord &>(rr)));
+        }
+    }
+
+    // Calculate the lease renew time based on update message tx time
+    // and the lease time. `kLeaseRenewGuardInterval` is used to
+    // ensure that we renew the lease before server expires it. In the
+    // unlikely (but maybe useful for testing) case where the accepted
+    // lease interval is too short (shorter than the guard time) we
+    // just use half of the accepted lease interval.
+
+    if (mAcceptedLeaseInterval > kLeaseRenewGuardInterval)
+    {
+        mLeaseRenewTime += Time::SecToMsec(mAcceptedLeaseInterval - kLeaseRenewGuardInterval);
+    }
+    else
+    {
+        mLeaseRenewTime += Time::SecToMsec(mAcceptedLeaseInterval) / 2;
+    }
+
+    // State changes:
+    //   kAdding     -> kRegistered
+    //   kRefreshing -> kRegistered
+    //   kRemoving   -> kRemoved
+
+    mHostInfo.SetState(kNewStateOnUpdateDone[mHostInfo.GetState()]);
+
+    for (Service *service = mServices.GetHead(); service != nullptr; service = service->GetNext())
+    {
+        ItemState oldState = service->GetState();
+
+        service->SetState(kNewStateOnUpdateDone[oldState]);
+
+        if (service->GetState() != oldState)
+        {
+            service->SetLeaseRenewTime(mLeaseRenewTime);
+        }
+    }
+
+    HandleUpdateDone();
+
+    UpdateState();
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        otLogInfoSrp("[clinet] Failed to process response %s", otThreadErrorToString(error));
+    }
+}
+
+void Client::HandleUpdateDone(void)
+{
+    HostInfo            hostInfoCopy = mHostInfo;
+    LinkedList<Service> removedServices;
+
+    if (mHostInfo.GetState() == kRemoved)
+    {
+        mHostInfo.Clear();
+    }
+
+    ResetRetryWaitInterval();
+    SetState(kStateUpdated);
+
+    GetRemovedServices(removedServices);
+    InvokeCallback(OT_ERROR_NONE, hostInfoCopy, removedServices.GetHead());
```

Comment:
Are we sure that `removedServices.GetHead() != nullptr` here?

Reply:
It is perfectly fine for it to be `nullptr`. It shows that no service was removed.

The `otSrpClientCallback` documentation mentions this.

```
 * @param[in] aService          The head of linked-list containing all services (excluding the ones removed). NULL if
 *                              the list is empty.
 * @param[in] aRemovedServices  The head of linked-list containing all removed services. NULL if the list is empty.
 * @param[in] aContext          A pointer to an arbitrary context (provided when callback was registered).
 *
 */
typedef void (*otSrpClientCallback)(otError                    aError,
                                    const otSrpClientHostInfo *aHostInfo,
                                    const otSrpClientService * aServices,
                                    const otSrpClientService * aRemovedServices,
                                    void *                     aContext);
```